src/core/cigar.rs

```diff
@@ -40,30 +40,19 @@ fn validate_code(code: &str) -> Result<()> {
 }
 
 impl Cigar {
-    pub fn new_with_code_and_raw(verfer: &Verfer, code: &str, raw: &[u8]) -> Result<Cigar> {
-        validate_code(code)?;
-        let mut cigar: Cigar = Matter::new_with_code_and_raw(code, raw)?;
-        cigar.set_verfer(verfer);
-        Ok(cigar)
-    }
-
-    pub fn new_with_qb64(verfer: &Verfer, qb64: &str) -> Result<Cigar> {
-        let mut cigar: Cigar = Matter::new_with_qb64(qb64)?;
-        cigar.set_verfer(verfer);
-        validate_code(&cigar.code())?;
-        Ok(cigar)
-    }
-
-    pub fn new_with_qb64b(verfer: &Verfer, qb64b: &[u8]) -> Result<Cigar> {
-        let mut cigar: Cigar = Matter::new_with_qb64b(qb64b)?;
-        cigar.set_verfer(verfer);
-        validate_code(&cigar.code())?;
-        Ok(cigar)
-    }
-
-    pub fn new_with_qb2(verfer: &Verfer, qb2: &[u8]) -> Result<Cigar> {
-        let mut cigar: Cigar = Matter::new_with_qb2(qb2)?;
-        cigar.set_verfer(verfer);
+    pub fn new(
+        verfer: Option<&Verfer>,
+        code: Option<&str>,
+        raw: Option<&[u8]>,
+        qb64b: Option<&mut Vec<u8>>,
```

Comment:
Are you confident about using mutable references to vectors, in the context of FFI?

Reply:
No, and I actually tried this out just now:

```rust
pub trait Uniffi {
    fn new(
        code: Option<String>,
        raw: Option<Vec<u8>>,
        qb64b: Option<Vec<u8>>,
        qb64: Option<String>,
        qb2: Option<Vec<u8>>,
    ) -> Result<Self> where Self: Sized;
}

impl Uniffi for Verfer {
    fn new(
        code: Option<String>,
        raw: Option<Vec<u8>>,
        qb64b: Option<Vec<u8>>,
        qb64: Option<String>,
        qb2: Option<Vec<u8>>,
    ) -> Result<Self> where Self: Sized {
        Verfer::create(
            code.map_or(None, |code| Some(&code)),
            raw.map_or(None, |raw| Some(&raw)),
            qb64b.map_or(None, |mut qb64b| Some(&mut qb64b)),
            qb64.map_or(None, |qb64| Some(&qb64)),
            qb2.map_or(None, |mut qb2| Some(&mut qb2)),
            &None
        )
    }
}
```
and renaming new to `create` for use in Rust, with the idea to include the trait from the uniffi lib.rs

It did not work.

We can drop the mut Vec, I was merely trying to accomodate the goal of parity - the python implementation does this.

Honestly I question using `uniffi`, I found pyo3 provided a more flexible interface in general but `uniffi` has more languages..

Reply:
I also think that it is rather `parside` responsibility to cut parsed bytes. `cesride` should provide a method to get the size of the parsed primitive.

Reply:
Also, it is easier to resize slices as it's just pointers to the collection.

Reply:
> I also think that it is rather parside responsibility to cut parsed bytes. cesride should provide a method to get the size of the parsed primitive.

This makes a lot of sense. So this is why you brought `full_size()` into your PR.

Reply:
When you said it's just pointers, obviously you'd still need to still call an alloc function under the hood to change the allocation on the heap - it isn't just changing the pointer. I'll check out what the difference is in rust. Can I just use indexing/slicing?

Reply:
It doesn't matter I suppose, if we move this functionality to parside.

Reply:
I tried this, I don't see what you mean and want to understand. When I use slicing like this on a &[u8]:
```rust
slice[length..];
```
I need to reassign that somehow, it doesn't mutate the variable in place. I don't see how it solves the problem.

Reply:
Also with `uniffi` we can't use &[u8] inside option. [ByRef] doesn't work that way as it is implemented. We also can't use mut afaict.

Reply:
> I tried this, I don't see what you mean and want to understand. When I use slicing like this on a &[u8]:
> 
> ```rust
> slice[length..];
> ```
> 
> I need to reassign that somehow, it doesn't mutate the variable in place. I don't see how it solves the problem.

I mean not mutating the input value but returning a new slice (which will have the same lifetime as the input) in the result. 
```
fn foo(bytes: &'a[u8]) -> Result<(Prefix, &'a[u8])>
```

Reply:
Ah, I see what you are saying. Yeah, I mean if we are starting to enter this territory of more complex return types, potentially through the FFI, I definitely think we should be doing all this in `parside` with something like `full_size()`